libkbfs/file_data.go

```diff
@@ -97,6 +122,341 @@ func (fd *fileData) getFileBlockAtOffset(ctx context.Context,
+			// amount the children.
+			nextBlockOff = iptr.Off
+			break
+		}
+
+		ptr := iptr.BlockPointer
+		childIndex := i
+		respCh := make(chan resp, 1)
+		respChans = append(respChans, respCh)
+		eg.Go(func() error {
+			block, err := fd.goGetter(groupCtx, fd.kmd, ptr, fd.file)
+			if err != nil {
+				return err
+			}
+			// Recurse down to the level of the child.
+			pfr, blocks, nextBlockOff, err := fd.getLeafBlocksForOffsetRange(
+				groupCtx, block, startOff, endOff, prefixOk)
+			if err != nil {
+				return err
+			}
+
+			// Append self to the front of every path and add `block` to
+			// the blocks map.
+			var r resp
+			for _, p := range pfr {
+				newPath := append([]parentBlockAndChildIndex{{
+					pblock:     pblock,
+					childIndex: childIndex,
+				}}, p...)
+				r.pathsFromRoot = append(r.pathsFromRoot, newPath)
+			}
+			r.blocks = make(map[BlockPointer]*FileBlock)
+			r.blocks[ptr] = block
+			for ptr, block := range blocks {
+				r.blocks[ptr] = block
+			}
+			r.nextBlockOff = nextBlockOff
+			respCh <- r
+			return nil
+		})
+	}
+
+	// If we are ok with just getting the prefix, don't treat a
+	// deadline exceeded error as fatal.
+	if err := eg.Wait(); err != nil && (!prefixOk ||
+		err != context.DeadlineExceeded) {
+		return nil, nil, 0, err
+	}
+
+	blocks = make(map[BlockPointer]*FileBlock)
+outer:
+	for i, respCh := range respChans {
+		select {
+		case r := <-respCh:
+			pathsFromRoot = append(pathsFromRoot, r.pathsFromRoot...)
+			for ptr, block := range r.blocks {
+				blocks[ptr] = block
+			}
+			if i == len(respChans)-1 {
+				// Figure out the offset of the next data block after
+				// the last in-range one.  It's the minimum of the
+				// next one at this level and the next one at the
+				// child's level, not counting -1 (which means we are
+				// returning the the last block at the given level).
+				if (r.nextBlockOff != -1 && r.nextBlockOff < nextBlockOff) ||
+					nextBlockOff == -1 {
+					nextBlockOff = r.nextBlockOff
+				}
+			}
+		default:
+			// There should always be a response ready in every
+			// channel, unless prefixOk is true.
+			if prefixOk {
+				break outer
+			} else {
+				panic("No response ready when !prefixOk")
+			}
+		}
+	}
+	return pathsFromRoot, blocks, nextBlockOff, nil
+}
+
+// getByteSlicesInOffsetRange returns an ordered, continuous slice of
+// byte ranges for the data described by the half-inclusive offset
+// range `[startOff, endOff)`.  If `endOff` == -1, it returns data to
+// the end of the file.  The caller is responsible for concatenating
+// the data into a single buffer if desired. If `prefixOk` is true,
+// the function will ignore context deadline errors and return
+// whatever prefix of the data it could fetch within the deadine.
+func (fd *fileData) getByteSlicesInOffsetRange(ctx context.Context,
+	startOff, endOff int64, prefixOk bool) ([][]byte, error) {
+	if (endOff != -1 && endOff <= startOff) || startOff < 0 {
```

Comment:
Hmm shouldn't it be something like:

```
if startOff < 0 || endOff < -1 {
  return nil, fmt.Errorf("Bad offset ...")
} else if endOff != -1 && endOff <= startOff {
  return nil, nil
}
```

?

The current logic returns no error when `startOff == -1`